### PR TITLE
Groups/samples_count

### PR DESCRIPTION
**Pull request by malchua**
## What does this PR do and why?
_Describe in detail what your merge request does and why._

This PR adds a samples_count column to Groups to show the total number of samples contained within a group. This also includes all samples in projects within subgroups.

## Screenshots or screen recordings
_Screenshots are required for UI changes, and strongly recommended for all other pull requests._
image

## How to set up and validate locally
_Numbered steps to set up and validate the change are strongly suggested._
1. Sign in
2. Click on "Groups" on the sidebar
3. Find a group where you are the owner and take note of the number of samples listed (number next to the beaker icon)
4. Select the group
5. Look through "Subgroups and projects" and check that the number of samples is accumulating correctly as you move up from projects to subgroups or subgroups to the parent group
6. Compare the number of samples listed in step 3 and see that it equals the accumulated number from step 5
7. Select a project within that group
8. Click on "Samples" on the sidebar
9. Add a new sample to the project
10. Go back to the parent group and check to see that the number of samples increased by one
11. Repeat steps 8-10 but delete samples
12. Repeat steps 8-10 but clone samples
13. Repeat steps 8-10 but transfer samples
14. Click on "Projects" on the sidebar
15. Select a project where you are the owner
16. Select "Settings" -> "General" in the sidebar
17. Go to the Transfer project section and transfer the project to a different parent group
18. Click on "Projects" on the sidebar
19. Find the parent group and check that the sample counts are correct.
20. Go back into the same project and delete the project
21. Click on "Projects" on the sidebar
22. Find the parent group of the deleted project
23. Check that the sample counts are correct
24. Repeat steps 14-23 but transfer and delete a group instead

## PR acceptance checklist
This checklist encourages us to confirm any changes have been analyzed to reduce risks in quality, performance, reliability, security, and maintainability.

- [x] I have evaluated the [PR acceptance checklist](https://phac-nml.github.io/irida-next/docs/development/development_processes/code_review#acceptance-checklist) for this PR.

### CODE REVIEW

File: db/migrate/20240925230422_add_samples_count_to_namespaces.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
+# migration to add samples count to namespaces table
+class AddSamplesCountToNamespaces < ActiveRecord::Migration[7.2]
+  def change
+    add_column :namespaces, :samples_count, :integer, default: 0
+  end
+end
```
Comment (ChrisHuynh333) [inline db/migrate/20240925230422_add_samples_count_to_namespaces.rb:8]: Much like how you had a migration to reset the sample counter for projects after adding the `samples_count` column to projects, you'll want a migration to set the current sample count of groups. If I were to seed my db on main, and then go to your branch and run the migration, my group sample counts are zero.

**Review by ChrisHuynh333 — CHANGES_REQUESTED** [review]
A couple comments at first glance.

You'll want to add UI tests as well for each of the actions (ie: after deleting a sample, return to the group index page and verify the sample count in the group's table row is accurate).

**Comment by deepsidhu85** [comment]
@malchua I was testing out your PR and came across this:

image

The number of samples is not being displayed for the projects on the `Subgroups and projects` tab when a group is expanded

File: test/controllers/groups_controller_test.rb [hunk 2]
```diff
@@ -118,7 +118,7 @@ class GroupsControllerTest < ActionDispatch::IntegrationTest
   test 'should not delete a group' do
     sign_in users(:joan_doe)
 
-    group = groups(:group_one)
+    group = groups(:group_twelve)
```
Comment (ChrisHuynh333) [inline test/controllers/groups_controller_test.rb:121]: After updating `group.samples_count` this change shouldn't be necessary.

File: test/controllers/groups_controller_test.rb [hunk 3]
```diff
@@ -177,7 +177,7 @@ class GroupsControllerTest < ActionDispatch::IntegrationTest
 
   test 'should transfer group' do
     sign_in users(:john_doe)
-    group = groups(:group_one)
+    group = groups(:group_twelve)
```
Comment (ChrisHuynh333) [inline test/controllers/groups_controller_test.rb:180]: After updating `group.samples_count` this change shouldn't be necessary.

File: test/models/group_test.rb [hunk 4]
```diff
@@ -206,4 +209,48 @@ def setup
...
+    project31 = projects(:project31)
+    Project.reset_counters(project31.id, :samples_count)
+    project31.reload.samples_count
+
+    expected_samples_count = group12.samples_count
+    actual_samples_count = Project.joins(:namespace).where(namespace: { parent_id: group12.self_and_descendants })
+                                  .select(:samples_count).pluck(:samples_count).sum
+
+    assert_equal expected_samples_count, actual_samples_count
+  end
```
Comment (ChrisHuynh333) [inline test/models/group_test.rb:233]: Didn't realize you had started this before writing the top comment, so yes change this to the above in my opinion.

File: test/services/samples/destroy_service_test.rb [hunk 5]
```diff
@@ -126,5 +126,60 @@ def setup
...
+      # Reference group/projects descendants tree:
+      # group12 < subgroup12b (project30 > sample 33)
+      #    |
+      #    ---- < subgroup12a (project29 > sample 32) < subgroup12aa (project31 > sample34 + 35)
+      group12 = groups(:group_twelve)
+      subgroup12a = groups(:subgroup_twelve_a)
+      subgroup12b = groups(:subgroup_twelve_b)
+      subgroup12aa = groups(:subgroup_twelve_a_a)
+      project31 = projects(:project31)
+      sample34 = samples(:sample34)
```
Comment (ChrisHuynh333) [inline test/services/samples/destroy_service_test.rb:140]: Could you refactor these into this test file's setup and remove the individual setups as many tests use these fixtures. Feel free to look through the other test files you worked in and make the same change if they're called many times as well.

File: test/system/groups_test.rb [hunk 6]
```diff
@@ -278,15 +282,15 @@ def setup
...
     group3 = groups(:group_three)
     visit group_url(group1)
 
     click_on I18n.t('groups.sidebar.settings')
     click_link I18n.t('groups.sidebar.general')
 
     assert_selector 'h2', text: I18n.t('groups.edit.advanced.transfer.title')
-    within %(form[action="/group-1/transfer"]) do
+    within %(form[action="/group-2/transfer"]) do
       assert_selector 'input[type=submit]:disabled'
```
Comment (ChrisHuynh333) [inline test/system/groups_test.rb:294]: Undo these changes after adding `group.samples_count`

File: app/services/groups/destroy_service.rb [hunk 7]
```diff
@@ -8,6 +8,8 @@ class DestroyService < BaseService
     def initialize(group, user = nil, params = {})
       super(user, params.except(:group, :group_id))
       @group = group
+      @deleted_samples_count = Project.joins(:namespace).where(namespace: { parent_id: @group.self_and_descendants })
+                                      .select(:samples_count).pluck(:samples_count).sum
```
Comment (ChrisHuynh333) [inline app/services/groups/destroy_service.rb:12]: This should just be able to be `@deleted_samples_count = @group.samples_count`

**Review by ChrisHuynh333 — CHANGES_REQUESTED** [review]
@ericenns @deepsidhu85 Feel free to chime in if you guys don't feel the same about the following.

I think you'll have to add `group.samples_count` manually to all groups like you have for projects.

In addition, I'd add a test to each of group and project models to loop through all projects/groups and check the current fixture `samples_count` to the actual `samples_count` (retrieve the real samples_count from `Project.reset_counters` / your group scope). This will help keep everyone accountable to update the sample_counts as we continue to add samples to our fixtures and keep things accurate. Add a comment to each test with something like `If you've added samples to fixtures, update the sample_count to the corresponding project/group` as the project/group won't be specified when the test fails.

File: app/services/samples/transfer_service.rb [hunk 8]
```diff
@@ -77,6 +77,7 @@ def transfer(new_project_id, sample_ids) # rubocop:disable Metrics/MethodLength,
 
         @project.namespace.update_metadata_summary_by_sample_transfer(transferred_samples_ids,
                                                                       new_project_id)
+        update_samples_count(transferred_samples_ids.count) if @project.parent.type == 'Group'
```
Comment (ericenns) [inline app/services/samples/transfer_service.rb:80]: **nitpick (blocking):** If I create a project under my user namespace with a sample, and then transfer that sample to a project that is under a group. The group does not update the sample count and it remains at 0.

**Review by ericenns — CHANGES_REQUESTED** [review]
I found a few edge cases where samples_count is not properly updated.

**Review by ChrisHuynh333 — APPROVED** [review]
Looks good!

**Review by ericenns — APPROVED** [review]
Looks good to me!

**Pull request merged**